### PR TITLE
scide: make tooltip text black on Linux

### LINKED ISSUE #1704: Autocomplete tooltip colors are unreadable on Linux
The fonto color of the tooltip is white over yellow which makes it not readable.

white-yellow

Ubuntu 14.04 Qt from Digia and Ubuntu.

### PR DESCRIPTION
**Pull request by snappizz**
I have no clue what I did here but it seems to fix #1704.

### CODE REVIEW

**Comment by bagong** [comment]
Hmm, this would indicate that the Q_WS_X11 branch is not used in a Linux compile, and in fact stackoverflow show various topics (but nothing very clear) that Q_WS_* should be changed to Q_OS_* when switching from Qt4 to Qt5. I think that makes sense with when that error turned up. See for example here:

https://www.kdab.com/porting-from-qt-4-to-qt-5/

So rather than changing the Mac/Windows code to Linux code, it might be worth trying if Q_OS_LINUX does the job?

**Comment by snappizz** [comment]
Aha, hopefully we'll manage to kill several birds with one stone here. Unfortunately, the code depends on [QGtkStyle](https://github.com/supercollider/supercollider/blob/master/editors/sc-ide/widgets/code_editor/autocompleter.cpp#L44), which seems to no longer exist in Qt 5.

**Comment by snappizz** [comment]
I've force pushed a commit that removes the dependency on QGtkStyle and just hardcodes the colors on Linux.

**Comment by bagong** [comment]
Maybe wise:
https://bugzilla.redhat.com/show_bug.cgi?id=1386404
So what do we have here? Window-manager vs. OS, Qt-company shaking of competitors, distribution policy, Linux flavor differences... Just a little preprocessor macro, and composers become engineers. We were hoping Qt would make things easier ;)
Nevertheless this looks really promising. I think it will be the favorite fix in 3.9!

**Comment by bagong** [comment]
Hah, it's the only spot where QGtKStyle is used in the entire app. That looks like a worthy sacrifice until we get a Qt crack...

But: UNIX includes MACOS. Is that intended (I checked, if I change the color in the Q_OS_UNIX branch, the color changes on the mac too)?

**Comment by snappizz** [comment]
OK, `Q_OS_LINUX` it is. This should be ready to merge after we've confirmed that the colors are correct on Linux, Mac, and Windows.

**Comment by bagong** [comment]
Really sorry for being nitpicking here, it would definitely be nice to get this out of the room. The only quick logic that seems safe here is UNIX but not (MACOS or netBSD). What is typically done actually is that things are defined for MACOS and Windows (together or separate) first, and then code for "the rest" (assuming that IOS is not an issue). I am not sure what LINUX covers (I used that in the beginning as a quick demonstrator, not as a final), and there are/used to be/could be quite a few unixes that are not covered by LINUX but could be covered by your "safe" code.

HAHA, for checking if the colors are right, we do not need to test. We'd have to use a different color to verify (and I did on OSX) ;) Btw, I'd like to see a unittest that catches the wrong color :smile:

**Comment by snappizz** [comment]
No worries about nitpicking. :heavy_check_mark:

**Comment by bagong** [comment]
Built with red on MacOS and Windows and got black ;) Thanks. A Qt-crack could maybe do something more nifty, but for our manpower I think this is an adequate fix and solves a tiny but annoying and much complained about problem ;)

It also can serve as model to fix some other problems at places using Q_WS_X11 which seem to have been overlooked when switching from Qt4 to Qt5. Those must all be going wrong, but they don't complain in the build as the alternative code compiles (but likely creates runtime bugs). I bet a bottle of wine that the keyboard problems on Linux have to do with this. ;)

**Comment by patrickdupuis** [comment]
I tested on my system and i get black. Should i be looking at text colour elsewhere?
tooltips-colour

**Comment by snappizz** [comment]
@patrickdupuis As I understand the issue, the problem is that the tooltip text color matches that of the window manager's text color, but the background color is fixed at pale yellow. Therefore, the text is only unreadable on systems where the native tooltips are light-on-dark.

**Comment by bagong** [comment]
@patrickdupuis What did you expect? It's hard-coded to black now...

**Comment by patrickdupuis** [comment]
Indeed it is. My system was never affected by this, and I can confirm that this change has no negative side effects for me.
 
Let's merge and make our Ubuntu users happy. Favourite fix of 3.9 for sure ;)

**Comment by bagong** [comment]
3.9 can come :smile:

**Comment by brianlheim** [comment]
Waiting for @snappizz to add a comment about the #if as requested, and then I'll be happy to merge.

**Comment by snappizz** [comment]
@brianlheim done, thanks for the reminder

**Comment by brianlheim** [comment]
Woah. Very thorough, thanks @snappizz 👍

File: editors/sc-ide/widgets/code_editor/autocompleter.cpp [hunk 1]
```diff
@@ -72,24 +68,33 @@ class MethodCallWidget : public QWidget
...
+// vestigial, but we REALLY needed to get this fix through since Linux
+// users have had unreadable autocomplete widgets for almost 18 months
+// now. We figured it was best to leave it alone.
+
+// See: https://github.com/supercollider/supercollider/pull/2762
+
+#if defined(Q_OS_UNIX) && !defined(Q_OS_MAC) && !defined(__NetBSD__)
+        QPalette p;
+        p.setColor( QPalette::Window, QColor(255, 255, 220) );
+        p.setColor( QPalette::WindowText, Qt::black );
```
Comment (timblechmann) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: this will break dark color schemes

Comment (bagong) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: Break them or have an unfavorable appearance in them? We are not able to implement configurability and theme support tooltips on linux. If you could provide it, that would be highly welcome! The current state is unacceptable for a lot of Linux users.

Comment (bagong) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: This is how it looks in Ubuntu Studio (xfce) with your dark scheme:
tooltip

Comment (snappizz) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: If someone would step up and make the tooltip colors more attractive, I'd much appreciate it. I have no expertise in Qt and I was just fumbling around here to fix an issue that's been bugging users for almost a year and a half.

Comment (timblechmann) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: read the colors for the color roles and fix them up. hardcoding colors should be avoided. futhermore, if it is related to gnome/gtk, netbsd would suffer the same issue

Comment (snappizz) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: I agree that hardcoding colors is a bad idea, but I'm afraid to say I just can't dedicate the time to finding a better solution.

Comment (patrickdupuis) [inline editors/sc-ide/widgets/code_editor/autocompleter.cpp]: We would need to compare the QColor::lightness() of both QPalette::ToolTipBase and QPalette::ToolTipText, and if they are unacceptably close, QColor::darker() the lighest one and QColor::lighter the darkest one by a significant factor?

**Comment by smoge** [comment]
I think uncontroversial commits like this one should be merged back into the last stable branch (3.8 in this case), where users can get the version with less bugs. Maybe a policy to follow?

**Pull request merged**